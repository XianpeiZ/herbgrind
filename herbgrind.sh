#!/usr/bin/env bash

# exit on error
set -e

# determine physical directory of this script
src="${BASH_SOURCE[0]}"
while [ -L "$src" ]; do
  dir="$(cd -P "$(dirname "$src")" && pwd)"
  src="$(readlink "$src")"
  [[ $src != /* ]] && src="$dir/$src"
done
MYDIR="$(cd -P "$(dirname "$src")" && pwd)"

HG="$MYDIR/valgrind/herbgrind-install/bin/valgrind --tool=herbgrind"

$HG "$@"
