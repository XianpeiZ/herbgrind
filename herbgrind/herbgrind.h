#ifndef __HERBGRIND_H
#define __HERBGRIND_H

#include "../include/valgrind.h"

// Let's define some user requests!
typedef enum {
  // This will start our tools instrumentation
  VG_USERREQ__BEGIN,
  // And this will end it.
  VG_USERREQ__END
} Vg_HerbGrindClientRequests;

#define HERBGRIND_BEGIN() \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0 /* default return (What does that mean?) */, \
                                 VG_USERREQ__BEGIN, \
                                 0, 0, 0, 0, 0); \
      _qzz_res; \
    }))

#define HERBGRIND_END() \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0, \
                                 VG_USERREQ__END, \
                                 0, 0, 0, 0, 0); \
      _qzz_res; \
    }))

#endif
