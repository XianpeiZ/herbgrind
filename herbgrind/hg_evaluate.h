#ifndef _HG_EVALUATE
#define _HG_EVALUATE

#include "hg_types.h"
#include "pub_tool_libcprint.h"

// Get the number of floating point values between x and y, including
// y.
unsigned long long ulpd(double x, double y);

// Evaluate the error of an executed operation.
void evaluateOpError(ShadowValue* shadowVal, double actualVal);


#endif
