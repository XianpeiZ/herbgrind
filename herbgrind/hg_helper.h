#ifndef _HG_HELPER
#define _HG_HELPER

#include "pub_tool_basics.h"

const char* IRExprTagString(IRExprTag tag);
const char* IRExprTagString(IRExprTag tag){
  switch(tag){
  case Iex_Binder:
    return "Iex_Binder";
  case Iex_Get:
    return "Iex_Get";
  case Iex_GetI:
    return "Iex_GetI";
  case Iex_RdTmp:
    return "Iex_RdTmp";
  case Iex_Qop:
    return "Iex_Qop";
  case Iex_Triop:
    return "Iex_Triop";
  case Iex_Binop:
    return "Iex_Binop";
  case Iex_Unop:
    return "Iex_Unop";
  case Iex_Load:
    return "Iex_Load";
  case Iex_Const:
    return "Iex_Const";
  case Iex_ITE:
    return "Iex_ITE";
  case Iex_CCall:
    return "Iex_CCall";
  case Iex_VECRET:
    return "Iex_VECRET";
  case Iex_BBPTR:
    return "Iex_BBPTR";
  }
  return "****";
}

#endif
