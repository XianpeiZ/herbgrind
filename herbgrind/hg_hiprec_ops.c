#include "hg_hiprec_ops.h"

int hiprec_yl2x(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_rnd_t round){
  mpfr_log2(result, arg2, round);
  return mpfr_mul(result, arg1, result, round);
}
int hiprec_yl2xp1(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_rnd_t round){
  mpfr_add_d(result, arg2, 1.0, round);
  mpfr_log2(result, result, round);
  return mpfr_mul(result, arg1, result, round);
}
int hiprec_scale(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_rnd_t round){
  mpfr_trunc(result, arg2);
  mpfr_exp2(result, result, round);
  return mpfr_mul(result, arg1, result, round);
}
