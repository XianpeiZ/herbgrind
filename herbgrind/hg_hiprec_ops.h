#ifndef _HG_HIPREC_OPS
#define _HG_HIPREC_OPS

#include "mpfr.h"

// These are just a bunch of composite mpfr operations to mirror some
// of the weird operations supported by the floating point unit. These
// operations will pretty much never do the smart thing, at least for
// now, so they may be the source of weird things down the line. mpfr
// operations return some sort of error code I think, here we're just
// returning the code from the last operation, and ignoring the
// others. There's no way that could possibly go wrong, right? right
// guys?

// One argument
int hiprec_2xm1(mpfr_t result, mpfr_t arg, mpfr_rnd_t round);
int hiprec_recpexp(mpfr_t result, mpfr_t arg, mpfr_rnd_t round);

// Two argument
int hiprec_yl2x(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_rnd_t round);
int hiprec_yl2xp1(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_rnd_t round);
int hiprec_scale(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_rnd_t round);

// Three argument
int hiprec_fma(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_t arg3, mpfr_rnd_t round);
int hiprec_fms(mpfr_t result, mpfr_t arg1, mpfr_t arg2, mpfr_t arg3, mpfr_rnd_t round);
#endif
