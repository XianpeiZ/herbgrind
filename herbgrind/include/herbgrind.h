
/*--------------------------------------------------------------------*/
/*--- HerbGrind: a valgrind tool for Herbie            herbgrind.h ---*/
/*--------------------------------------------------------------------*/

/*
   This file is part of HerbGrind, a valgrind tool for diagnosing
   floating point accuracy problems in binary programs and extracting
   problematic expressions.

   Copyright (C) 2016 Alex Sanchez-Stern

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 3 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
   02111-1307, USA.

   The GNU General Public License is contained in the file COPYING.
*/

#ifndef __HERBGRIND_H
#define __HERBGRIND_H

#include "../../include/valgrind.h"

// Let's define some user requests!
typedef enum {
  // This will start our tools instrumentation
  VG_USERREQ__BEGIN = VG_USERREQ_TOOL_BASE('H', 'B'),
  // And this will end it.
  VG_USERREQ__END,

  VG_USERREQ__PRINT,
  VG_USERREQ__PRINTNUM,

  VG_USERREQ__PERFORM_OP
} Vg_HerbGrindClientRequests;

#define HERBGRIND_BEGIN() \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0 /* default return (What does that mean?) */, \
                                 VG_USERREQ__BEGIN, \
                                 0, 0, 0, 0, 0); \
      _qzz_res; \
    }))

#define HERBGRIND_END() \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0, \
                                 VG_USERREQ__END, \
                                 0, 0, 0, 0, 0); \
      _qzz_res; \
    }))

#define HERBGRIND_PRINT(_qzz_mesg) \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0, \
                                 VG_USERREQ__PRINT, \
                                 _qzz_mesg, 0, 0, 0, 0); \
      _qzz_res; \
    }))

#define HERBGRIND_PRINTNUM(_qzz_num) \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0, \
                                 VG_USERREQ__PRINTNUM, \
                                 _qzz_num, 0, 0, 0, 0); \
      _qzz_res; \
    }))

#define HERBGRIND_PERFORM_OP(_qzz_op, _qzz_result_addr, _qzz_args)      \
  (__extension__({unsigned long _qzz_res; \
      VALGRIND_DO_CLIENT_REQUEST(_qzz_res, 0, \
                                 VG_USERREQ__PERFORM_OP, \
                                 _qzz_op, _qzz_result_addr, _qzz_args, 0, 0); \
      _qzz_res; \
    }))

#endif
