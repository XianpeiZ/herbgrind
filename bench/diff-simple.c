#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "herbgrind.h"

int main(int argc, char** argv){
  double x,y;
  x = 1e16;
  y = (x + 1) - x;
  printf("%e\n", y);
  return 0;
}
