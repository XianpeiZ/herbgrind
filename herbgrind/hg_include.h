#ifndef _HG_INCLUDE
#define _HG_INCLUDE

#include "pub_tool_basics.h"
#include "pub_tool_tooliface.h"

// For handling client requests. Those are those macros that you can
// put in your code to be instrumented, which will allow you to talk
// to the tool. For instance, HERBGRIND_BEGIN() and HERBGRIND_END()
// will turn the tool on an off for certain sections.
#include "herbgrind.h"

// This file has all the functions and data structures that will be
// called by the instrumented program.
#include "hg_runtime.h"
// This file has the code to instrument each statement in the client
// program.
#include "hg_instrument.h"

// Pull in this header file so that we can call the valgrind version
// of printf.
#include "pub_tool_libcprint.h"
// Pull in this header file so that we can set the strlen, strcpy,
// memmove, memcmp, and memset functions of mpfr to their valgrind
// library equivalents.
#include "pub_tool_libcbase.h"

// This does any initialization that needs to be done after command
// line processing.
static void hg_post_clo_init(void);
// This is where the magic happens. This function gets called to
// instrument every superblock.
static IRSB* hg_instrument ( VgCallbackClosure* closure,
                             IRSB* bb,
                             const VexGuestLayout* layout, 
                             const VexGuestExtents* vge,
                             const VexArchInfo* archinfo_host,
                             IRType gWordTy, IRType hWordTy );
// This is called after the program exits, for cleanup and such.
static void hg_fini(Int exitcode);
// This is called after the program exits, for cleanup and such.
// This handles client requests, the macros that client programs stick
// in to send messages to the tool.
static Bool hg_handle_client_request(ThreadId tid, UWord* arg, UWord* ret);
// This is where we initialize everything
static void hg_pre_clo_init(void);

// Print an IR superblock
static void printSuperBlock(IRSB* superblock);

//#define PRINTBLOCKS

#endif
