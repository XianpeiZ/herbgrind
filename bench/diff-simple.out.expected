(output
  (function "main")
  (filename "diff-simple.c")
  (line-num 10)
  (instr-addr 400829)
  (avg-error 61.998590)
  (max-error 61.998590)
  (num-calls 1)
  (influences
    (
    (
     (expr
       (FPCore ()
          (- (+ 1.000000 1.000000e16) 1.000000e16)))
     (var-problematic-ranges)
     (example problematic input ())
     (function "main")
     (filename "diff-simple.c")
     (line-num 9)
     (instr-addr 400804)
     (avg-error 61.998590)
     (max-error 61.998590)
     (avg-local-error 61.998590)
     (max-local-error 61.998590)
     (num-calls 1))
    )
  )
)
